src/pubsub.rs

```diff
@@ -0,0 +1,215 @@
+use std::collections::HashMap;
+
+use abscissa_core::{
+    tracing::log::{debug, error},
+    Application,
+};
+use somm_proto::{
+    cosmos_sdk_proto::cosmos::base::abci::v1beta1::TxResponse,
+    pubsub::{
+        query_client::QueryClient, QueryDefaultSubscriptionsRequest,
+        QueryDefaultSubscriptionsResponse, QueryPublisherRequest, QueryPublisherResponse,
+        QuerySubscriberIntentsBySubscriberAddressRequest,
+        QuerySubscriberIntentsBySubscriberAddressResponse, Subscriber,
+    },
+};
+use url::Url;
+use x509_parser::prelude::parse_x509_pem;
+
+use crate::{
+    config::get_delegate_address,
+    error::{Error, ErrorKind},
+    prelude::APP,
+    somm_send,
+};
+
+use self::cache::PublisherTrustData;
+
+pub(crate) mod cache;
+
+/// Retrieves the PEM encoded CA certs for all default subscription publishers.
+pub(crate) async fn get_trust_state() -> Result<Vec<PublisherTrustData<'static>>, Error> {
+    debug!("collecting trust roots from default subscriptions and subscriber intents.");
+
+    // load subscription mappings
+    let mut subscription_mappings: HashMap<String, Vec<String>> = HashMap::default();
+    get_default_subscriptions()
+        .await?
+        .default_subscriptions
+        .iter()
+        .for_each(|s| {
+            subscription_mappings
+                // if there is an entry for the key, push to the vec and dedup,
+                // otherwise create a new vec with the subscription id
+                .entry(s.publisher_domain.to_owned())
+                .and_modify(|e| {
+                    e.push(s.subscription_id.to_owned());
+                    e.dedup();
+                })
+                .or_insert(vec![s.subscription_id.to_owned()]);
+        });
+    let delegate_address = get_delegate_address().to_string();
+    get_subscriber_intents_by_subscriber_address(delegate_address)
+        .await?
+        .subscriber_intents
+        .iter()
+        .for_each(|si| {
+            subscription_mappings
+                .entry(si.publisher_domain.to_owned())
+                .and_modify(|e| {
+                    e.push(si.subscription_id.to_owned());
```

Comment:
I think the issue here is that subscriber intents must override default subscriptions. Which means that first collecting the default subscriptions and push/dedup, and then doing the same for the subscriber intents on the same map, would have the effect of duplicating a subscription to two different publishers if the subscriber had overridden the default value.

Reply:
Ah good catch.